Mathlib/RingTheory/AdicCompletion/Basic.lean

```diff
@@ -83,18 +83,27 @@ def Hausdorffification : Type _ :=
   M ⧸ (⨅ n : ℕ, I ^ n • ⊤ : Submodule R M)
 #align Hausdorffification Hausdorffification
 
-/-- The completion of a module with respect to an ideal. This is not necessarily Hausdorff.
-In fact, this is only complete if the ideal is finitely generated. -/
-def adicCompletion : Submodule R (∀ n : ℕ, M ⧸ (I ^ n • ⊤ : Submodule R M)) where
-  carrier := { f | ∀ {m n} (h : m ≤ n), liftQ _ (mkQ _) (by
-      rw [ker_mkQ]
-      exact smul_mono (Ideal.pow_le_pow_right h) le_rfl)
-    (f n) = f m }
+/-- The canonical linear map `M ⧸ (I ^ n • ⊤) →ₗ[R] M ⧸ (I ^ m • ⊤)` for `m ≤ n` used
+to define `AdicCompletion`. -/
+def AdicCompletion.transitionMap {m n : ℕ} (hmn : m ≤ n) :
+    M ⧸ (I ^ n • ⊤ : Submodule R M) →ₗ[R] M ⧸ (I ^ m • ⊤ : Submodule R M) :=
+  liftQ (I ^ n • ⊤ : Submodule R M) (mkQ (I ^ m • ⊤ : Submodule R M)) (by
+    rw [ker_mkQ]
+    exact smul_mono (Ideal.pow_le_pow_right hmn) le_rfl)
+
+/-- Auxiliary definition for `AdicCompletion`. -/
+private def adicCompletion : Submodule R (∀ n : ℕ, M ⧸ (I ^ n • ⊤ : Submodule R M)) where
+  carrier := { f | ∀ {m n} (hmn : m ≤ n), AdicCompletion.transitionMap I M hmn (f n) = f m }
   zero_mem' hmn := by rw [Pi.zero_apply, Pi.zero_apply, LinearMap.map_zero]
   add_mem' hf hg m n hmn := by
     rw [Pi.add_apply, Pi.add_apply, LinearMap.map_add, hf hmn, hg hmn]
   smul_mem' c f hf m n hmn := by rw [Pi.smul_apply, Pi.smul_apply, LinearMap.map_smul, hf hmn]
-#align adic_completion adicCompletion
+
+/-- The completion of a module with respect to an ideal. This is not necessarily Hausdorff.
+In fact, this is only complete if the ideal is finitely generated. -/
+def AdicCompletion : Type _ :=
+  adicCompletion I M
```

Comment:
Maybe we can define this to be `{ f // ∀ {m n} (hmn : m ≤ n), AdicCompletion.transitionMap I M hmn (f n) = f m }` and ditch `adicCompletion` entirely? Similarly for `adicCauchySequence` (maybe replace it with a predicate `IsAdicCauchy`)

Reply:
How do I then get the `AddCommGroup` and `Module` instances? Do I have to write them by hand? Currently they are inherited by `adicCompletion` being a `Submodule`.

Reply:
And I have realized that `adicCompletion` should not be private after all, since it comes in handy again for defining the algebra instance. Maybe rename `adicCompletion` in `submodule` and define `AdicCompletion` as you suggest. Then I think `inferInstanceAs <| AddCommGroup (submodule I M)` should still work.

Reply:
I suppose the proof that pointwise add and smul works is exactly the same proof as showing `adicCompletion` is a submodule. But if you will need it elsewhere then I think your suggestion is the way to go.

Reply:
I think I changed it according to your suggestions.